<a href="https://colab.research.google.com/github/lrpreuss/inserir_arquivos_aulas_LinguagemProgramacaoI/blob/main/Aula05_Fundamentos_de_Python_0904.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Conteúdo do dia

##Caderno

Manipulação de arquivos ~ via input (entrada de dados) = via usuário, linhas de comando, arquivos. Hoje: como processar arquivos .txt.

Com Python, é importante saber como o sistema das informações é organizado.
* Referência do arquivo. A ideia é basicamente saber o caminho do arquivo. Existe a referência **absoluta** e a **relativa**. Por vezes, a absoluta (caminho todo) pode ser 'comprida', mas você não terá problemas com o caminho. A outra opção é a referência relativa, que vai from arquivo python até o arquivo (como quando o arquivo a ser consultado está salva mesma pasta do projeto). Ao abrirmos arquivos em projetos python, é necessário **referenciar** esses arquivos.

É bom que o arquivo sempre esteja **dentro de uma variável** que facilita a manipulação dos dados. Esse processo de manipulação sempre começa com a criação de um OBJETO DE ARQUIVO (como se fosse uma 'cópia' que só o script usa para trabalhar, sem alterar o arquivo original.)

* Ciclo de vida de um arquivo: abertura, manipulação (leitura/ gravação) e fechamento
```open('caminho do arquivo', modo que será aberto (leitura/ escrita/ etc), codificação)```
Por exemplo: ```arquivo_teste = open('teste.txt'). ```
Caso o arquivo não exista, vai dar erro *not found*.


* Modos possíveis: r = leitura // w = escrita // a = adição ao final do texto // b = arquivo binário // t = texto (a codificação só é aplicada aos arquivos de texto, pra ele rodar para linguagem natural do jeito certo. Por padrão, o python reconhece tudo como arquivo de texto na codificação utf-8.)



[traceback: a sintaxe está certa, mas existe outro erro no processo - tipo quando um arquvo não existe.]

**Métodos**: funções aplicadas a determinados tipos de objeto. Exemplo:
Texto.read() = objeto arquivo
format() = objetos string.
Há funções só para listas também, entre outros.
.read() = leitura total
.readline = lê uma linha
.seek(posição) = posição desejada para o cursor
.close() = fecha o arquivo
.write() = escreve no arquivo


## EXEMPLOS

In [ ]:
#exemplo:

variavel = 'caminho do arquivo'
texto = open(variavel, 'r', encoding='utf-8')
#leitura(read)
palavras = texto.read()
texto.close()

In [ ]:
## criação da função: abertura, manipulação e fechamento

#Opção1
def leituraArquivo(nomeDoArquivo):
  """descrição do que a função faz logo na primeira linha
  várias linhas de comentário"""

  arqEntrada = open(nomeDoArquivo, 'r')
  conteudo = arqEntrada.read()
  arqEntrada.close()

  return conteudo

#Opção2
def leituraArquivo(nomeDoArquivo):
  """descrição do que a função faz logo na primeira linha
  várias linhas de comentário"""

  arqEntrada = open(nomeDoArquivo, 'r')
  conteudo = arqEntrada.read()
  arqEntrada.close()

  return len(conteudo) #informa o tamanho do arquivo

#Opção3
def leituraArquivo(nomeDoArquivo):
  """descrição do que a função faz logo na primeira linha
  várias linhas de comentário"""

  arqEntrada = open(nomeDoArquivo, 'r')
  conteudo = arqEntrada.read()
  ListaPalavras = conteudo.split()  #quando o texto é muito grande, não é recomendado trabalhar como um bloco só - então transformamos em uma lista de palavras.
  #Isso é interessante quando se precisa igorar certas palavras também.
  arqEntrada.close()

  return ListaPalavras

## EXERCÍCIO 23 - Lendo arquivos

In [1]:
from sys import argv #do sistema, pegou a biblioteca argv

script, filename = argv #o que vamos usar do argv
txt = open(filename) #quando chamar o txt, vai abrir conforme o nome dado

print(f"Here's yout file {filename}:") #imprime a frase + nome do arquivo
print(txt.read()) #exibe o que tem no arquivo txt

print("Type the filename again:") #imprime a frase
file_again = input("> ") #input para file_again

txt_again = open(file_again) #vai abrir o que foi definido como txt_again

print(txt_again.read()) #imprime o que leu de txt_again

ValueError: too many values to unpack (expected 2)

### Novidades neste exercício

* Testes para abrir arquivos .txt direto no código.

### Erros que cometi neste exercício

"script, filename = argv #o que vamos usar do argv" -> ValueError: too many values to unpack (expected 2)

## EXERCÍCIO 24 - Lendo e gravando arquivos

In [ ]:
from sys import argv

scipt, filename = argv

print(f"We're going to erase {filename}.")
print("If you don't want that, hit CTRL-C (^C).")
print("If you do want that, hit RETURN.")

input("?")

print("Opening the file...")
target = open(filename, 'w')

print("Intruncating the file. Goodbye!")
target.truncate()

print("Now I'm going to ask you for three lines.")

line1 = input("line 1: ")
line2 = input("line 2: ")
line3 = input("line 3: ")

print("I'm going to write these to the file.")

target.write(line1)
target.write("\n")
target.write(line2)
target.write("\n")
target.write(line3)
target.write("\n")

print("And finally, we close it.")
target.close

### Novidades neste exercício

* Escrever e gravar coisas em .txt direto no código.

### Erros que cometi neste exercício

Não aprendi a fazer isso funcionar e já tô frustrada demais pra continuar. Outra hora tento de novo.

## EXERCÍCIO 25 - Mais Arquivos

In [ ]:
from sys import argv
from os.path import exists

script, from_file, to_fle = argv

print(f"Copying from {from_file} to {to_file}")

#poderíamos fazer esses dois com uma linha, como?
in_file = open(from_file)
indata = in_file.read()

print(f"The input file is {len(indata)} bytes long")

print(f"Does the output file exist? {exists(to_file)}")
print("Ready, hit RETURN to contiue, CTRL - C to abort.")
input()

out_file = opento_file, 'w')
out_file.write(indata)

print("Alright, all done.")

out_file.close()
in_file.close()

### Novidades neste exercício

### Erros que cometi neste exercício

## EXERCÍCIO 26 - Funções e arquivos

In [ ]:
from sys import argv

script, input_file = argv

def print_all(f):
  print(f.read())

def rewind(f):
  f.seek(0)

def print_a_line(line_count, f):
  print(line_count, f.readline())

current_file = open(input_file)

print("First let's print the whole file:\n")

print_all(current_file)

print("Now let's rewind, kind of like a tape.")

rewind(current_file)

print("Let's print three lines:")

current_line = 1
print_a_line(current_line, current_file)

current_line + 1
print_a_line(current_line, current_file)

current_line + 1
print_a_line(current_line, current_file)

### Novidades neste exercício

* Abre o arquivo direto no código e permite uma busca dentro dele.

### Erros que cometi neste exercício

## EXERCÍCIO 27 - Mais prática

In [6]:
print("Let's practice everything.")
print('You\'d need to know \'bout escapes with \\ that do:')
print('\n newlines and \t tabs.')

poem = """
\tThe lovely world
with logic so firmly planted
cannot discern \n the needs of love
nor comprehend passion from intuition
and requires an explenation
\n\twhere there is none.
"""

print("---------------")
print(poem)
print("---------------")

five = 10 - 2 + 3 - 6
print(f"This should be five: {five}")

def secret_formula(started):
  jelly_beans = started * 500
  jars = jelly_beans / 1000
  crates = jars / 100
  return jelly_beans, jars, crates

start_point = 10000
jelly_beans, jars, crates = secret_formula(start_point) #não deveria ser jelly_beans? DEVERIA.

#lembre-se de que essa é uma outra maneira de formatar uma string
print("With a starting point of: {}".format(start_point))
#é comko usar a string f""
print(f"We'd have {jelly_beans} beans, {jars} jars, and {crates} crates.")

start_point = start_point / 10

print("We can also do this way:")
formula = secret_formula(start_point)
# essa é uma maneira fácil de aplicar uma lista a uma string de formatação
print("We'd have {} beans, {} jars, and {} crates.".format(*formula))

Let's practice everything.
You'd need to know 'bout escapes with \ that do:

 newlines and 	 tabs.
---------------

	The lovely world
with logic so firmly planted
cannot discern 
 the needs of love
nor comprehend passion from intuition
and requires an explenation

	where there is none.

---------------
This should be five: 5
With a starting point of: 10000
We'd have 5000000 beans, 5000.0 jars, and 50.0 crates.
We can also do this way:
We'd have 500000.0 beans, 500.0 jars, and 5.0 crates.


### Novidades neste exercício

* Relembra \n e \t
* Mostra a alternativa '.format' para 'f'

### Erros que cometi neste exercício

* jelly_beans virou beans e deu erro algumas vezes.
* o maior erro é a existência tornada inútil devido o acúmulo de capital por poucos às custas de muitos.

## EXERCÍCIO 28 - Desvios e Funções

In [8]:
from sys import exit

def sala_do_ouro():
  print("Esta sala está cheia de ouro. Quanto você pega?")
  print("Digite um número.")

  escolha = input("> ")
  if "0" in escolha or "1" in escolha:
    quanto = int(escolha)
  else:
    morrer("Cara, aprenda a digitar um número.")

  if quanto < 50:
    print("Legal, você não é ganancioso, você venceu!")
    exit(0)
  else:
    morrer("Seu ganancioso!")


def sala_do_urso():
  print("Há um urso aqui.")
  print("O urso tem um monte de mel.")
  print("O urso gordo está na frente de outra porta.")
  print("O que você faz?")
  print("Opções: 'pegar mel', 'provocar urso'.")
  urso_movido = False

  while True:
    escolha = input("> ")

    if escolha == "pegar mel":
      morrer("O urso olha para você e arranca sua cara com uma patada.")
    elif escolha == "provocar urso" and not urso_movido:
      print("O urso se moveu da porta.")
      print("Você pode 'abrir porta' agora.")
      urso_movido = True
    elif escolha == "provocar urso" and urso_movido:
      morrer("O urso fica puto e arranca sua perna.")
    elif escolha == "abrir porta" and urso_movido:
      sala_do_ouro()
    else:
      print("Não entendi sua escolha. Tente novamente.")


def sala_de_cthulhu():
  print("Aqui você vê o grande mal, Cthulhu.")
  print("Ele, isso, seja lá o que for, olha para você e você enlouquece.")
  print("O que você faz?")
  print("Opções: 'foge', 'cabeça'.")

  escolha = input("> ")

  if "foge" in escolha:
    comecar()
  elif "cabeça" in escolha:
    morrer("Bem, isso foi gostoso!")
  else:
    sala_de_cthulhu()


def morrer(porque): #esqueci o ':' na primeira rodada
  print(porque, "Bom trabalho!")
  return "The game is over"

def comecar():
  print("Você está em uma sala escura.")
  print("Há uma porta  à sua direita e à sua esquerda.")
  print("Qual você pega?")
  print("Opções: 'esquerda', 'direita'.")

  escolha = input("> ")

  if escolha == "esquerda":
    sala_do_urso()
  elif escolha == "direita":
    sala_de_cthulhu()
  else:
    morrer("Você tropeça pela sala até morrer de fome.")


comecar()

Você está em uma sala escura.
Há uma porta  à sua direita e à sua esquerda.
Qual você pega?
Opções: 'esquerda', 'direita'.
> direita
Aqui você vê o grande mal, Cthulhu.
Ele, isso, seja lá o que for, olha para você e você enlouquece.
O que você faz?
Opções: 'foge', 'cabeça'.
> foge
Você está em uma sala escura.
Há uma porta  à sua direita e à sua esquerda.
Qual você pega?
Opções: 'esquerda', 'direita'.
> esquerda
Há um urso aqui.
O urso tem um monte de mel.
O urso gordo está na frente de outra porta.
O que você faz?
Opções: 'pegar mel', 'provocar urso'.
> provocar urso
O urso se moveu da porta.
Você pode 'abrir porta' agora.
> abrir porta
Esta sala está cheia de ouro. Quanto você pega?
Digite um número.
> 1
Legal, você não é ganancioso, você venceu!


SystemExit: 0

/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py:3561: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)


### Novidades neste exercício

* Associação entre métodos, associações lógicas

### Erros que cometi neste exercício

* digitação;
* no final do código, ao responder '1':

An exception has occurred, use %tb to see the full traceback.

SystemExit: 0
/usr/local/lib/python3.11/dist-packages/IPython/core/interactiveshell.py:3561: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.
  warn("To exit: use 'exit', 'quit', or Ctrl-D.", stacklevel=1)
  
   -> não faço a menor ideia do que seja, ainda bem que a média é 6 e eu posso perder esse pontinho.